### ISSUE TITLE
FR: Keyboard moveable highlight (bounty to be set)

### ISSUE THREAD

**Issue by Ceynou** [koreader/koreader#12145]
Requires writing an entire plugin, I barely know how to code and... idk I guess I'm just not motivated to do it myself sorry:

Add an option in the Taps and Gestures menu to enable or disable the plugin. 
Start the highlight mode of the plugin when pressing the keyboard key "Press". 
Its initial "position" (the word it will highlight) can be approximated based on how long I was on that page compared to how long it takes me to read a page on average. 

That highlight should be moveable: 
Moving on each word of the current page with the keyboard keys "Left", "Right", "Up" and "Down", assuming we can get a get "Up" and "Down" working.

You can look the highlighted word up with the keyboard key "Press" when you're already in highlight mode, so it would bring up the default pop up dictionary.

You can exit and stop the highlight mode of the plugin either with the "Backspace" keyboard key or after looking up a word.

**Additional context**
That way I could use my 8BitDo without ever moving my hands, touching the screen to look up unknown words.
Tell me what the minimum amount of the bounty should be, and what method to use for the bounty (a site in particular?)

**Comment by Frenzie** [comment]
Actually this already exists, either nearly or entirely but I'm on my phone atm so I can't check. I think it may have still required some gamepad detection to be finalized for this specific use case.

For context, does this particular gamepad have a stick and a dpad or only a dpad? In the latter case things might be harder.

I don't think we have a bounty thing, nor that it would be very motivating one way or the other.

**Comment by Ceynou** [comment]
> Actually this already exists, either nearly or entirely but I'm on my phone atm so I can't check. I think it may have still required some gamepad detection to be finalized for this specific use case.
> 
> For context, does this particular gamepad have a stick and a dpad or only a dpad? In the latter case things might be harder.
> 
> I don't think we have a bounty thing, nor that it would be very motivating one way or the other.

The one I have only has a dpad (other models have a stick), it also has a keyboard mode I just need to bind the buttons to keyboard keys within the 8BitDo Ultimate application and it can act like any bluetooth keyboard device.

Is it harder because we're thinking about moving a cursor instead of a highlight?

**Comment by Frenzie** [comment]
> Is it harder because we're thinking about moving a cursor instead of a highlight?

No, but then the dpad isn't a bunch of free extra keys.

Relevant code here ([rationale here](https://github.com/koreader/koreader/pull/8914)):
https://github.com/koreader/koreader/blob/e29901828610f91897d2a0f444e489cd4bd83107/frontend/apps/reader/modules/readerhighlight.lua#L377-L384

https://github.com/koreader/koreader/blob/e29901828610f91897d2a0f444e489cd4bd83107/frontend/apps/reader/readerui.lua#L526-L528

**Comment by Ceynou** [comment]
Yeah but like I said the one I have (8BitDo micro) has a keyboard mode, I can bind any button to shift, Ctrl, qwerty, whatever key is in a keyboard, so I think we can ignore the gamepad support

**Comment by Ceynou** [comment]
Okay I tried the linux app, it's "h" to activate the highlight cursor.
I'll see if I can write the plugin with that, I'd rather have a moving highlight tho, that cross feels clunky tbh

**Comment by Frenzie** [comment]
Note that it has seen some minor tweaks since the last release.

**Comment by Commodore64user** [comment]
Is this what you have https://www.amazon.co.uk/dp/B0CDG5HCCH ?

I think you could get away with pretending to be a K4 when you connect the gamepad. That would reduce you workload to, faking being a K4 as the rest of the work is already done.

**Comment by Ceynou** [comment]
> Is this what you have https://www.amazon.co.uk/dp/B0CDG5HCCH ?

Yes this is what I use, I don't understand what you mean by K4 (kindle 4?), is it for the keybinds part of the plugin?

**Comment by Commodore64user** [comment]
yes, kindle 4. Essentially you want to give the gamepad some attributes (see below) so when it connects to KOReader you gain them, once you have those 5 things, you are a kindle 4 as far as the code is concerned.

```
isTouchDevice = false,
hasDPad = true,
useDPadAsActionKeys = true,
hasKeys = yes,
hasScreenKB = true,
```

I am not sure what else you would need, but I think this could make your adventure a lot easier than writing everything from scratch.

**Comment by Commodore64user** [comment]
hey @ryanwwest I recently read that you also have one of these gamepads. This is an idea that you could possibly use as well.

**Comment by Commodore64user** [comment]
we've done it, no longer clunky #15377 and upcoming gamepad support #15449. is the bounty still in the room with us? ;)

**Comment by Ceynou** [comment]
> we've done it, no longer clunky [#15377](https://github.com/koreader/koreader/pull/15377) and upcoming gamepad support [#15449](https://github.com/koreader/koreader/pull/15449). is the bounty still in the room with us? ;)

Yes it's still in the room, thank you for your hard work you guys are goated, I made a small donation because of my current situation

I installed the latest nightly build [apk](https://build.koreader.rocks/download/nightly/2026.03-169-gd246323ba/koreader-android-arm64-v2026.03-169-gd246323ba_2026-05-29.apk
) (koreader-android-arm64-v2026.03-169-gd246323ba_2026-05-29), I tried on my phone and on my ereader (BigMe B751), connected a keyboard by USB, I cannot move the indicator, with or without Shift, or any of combination of any modifier, what am I doing wrong?


EDIT: It does work on linux, when the indicator is activated, the dpad (arrow keys) does move the indicator instead of changing the page, and with Shift it does skip words or lines instead of of displaying whatever is assigned to these shortcuts on android(? bookmarks, ToC and book maps, shift + ~down~ right deactivates the indicator).
So I guess this is an issue with how android treats USB keyboard input?

**Comment by Commodore64user** [comment]
I am not familiar with Android, not sure what works and what doesn't, this is the only ticket #14358 I am aware of.

Are you saying that on Android, keyboard shortcuts work but using text selection does not? or are you saying that no keyboard input works at all? also `shift+down` should not stop the indicator, not sure what that is all about.

**Comment by Frenzie** [comment]
My gamepad PR only pertains to SDL. Hooking it up to Android is another matter, if presumably not too hard with the groundwork laid.

**Comment by Ceynou** [comment]
I'm on limited internet so I can't upload a video so I used this [website](https://w3c.github.io/uievents/tools/key-event-viewer.html) to compare my keyboard (Ducky One 2) inputs on linux (cachyos) and android (lineageos), I'm not using a gamepad and will probably never do:
<img width="1360" height="725" alt="Image" src="https://github.com/user-attachments/assets/2e24c482-1217-4ce8-9cda-a82fce6449b9" />

I pressed the same keys on android, for some reason the shift keypress isn't registered but it works as a modifier

<img src="https://github.com/user-attachments/assets/1eed48fc-5a25-4b15-b998-8e5204565415" width=40%> 

I have difficulties expressing myself, sorry if I sound cold, so the scenario is as follows:
Open a book on koreader, activate the text selection(?), move to the next and previous words, move to next and previous lines, skip words forward and backward, skip lines forward and backward, do something on the word.

**On linux:**
**h** --> activates the text selection and displays the crosshair on a word in the middle of the page
**left** --> moves crosshair to the previous word
**right** --> moves crosshair to the next word
**up** --> moves crosshair to the word on the previous line
**down** --> moves crosshair to the word on the next line

**shift+left** --> moves crosshair to the previous word skipping 3 words
**shift+right** --> moves crosshair to the next word skipping 3 words
**shift+up** --> moves crosshair to the word on the previous line skipping 3 lines
**shift+down** --> moves crosshair to the word on the next line skipping 3 lines

**enter** --> start selection, press enter again and  it displays the menu to choose what to do with the selection


**On android:**
**h** --> activates the text selection and displays the crosshair on a word in the middle of the page
**left** --> change to the previous page with the crosshair still activated
**right** --> change to the next page with the crosshair still activated
**up** --> does nothing
**down** --> does nothing

**shift+left** --> shows the Bookmarks menu
**shift+right** --> deactivate the text selection, crosshair disappear
**shift+up** --> shows the ToC menu
**shift+down** --> shows the Bookmap menu 

**enter** --> shows the bottom menu (Rotation and Two Columns)

**Comment by Commodore64user** [comment]
so, the keyboard works, just not for text selection... seems like the vents are not being registered, thus falling back to hotkeys/readerui events.

which version do you have?
try commenting out these lines and see what happens
https://github.com/koreader/koreader/blob/9c5e596098a8dd800db050a2df6e8a3f3a0a8dd6/frontend/apps/reader/modules/readerkeyselection.lua#L28-L30

**Comment by Ceynou** [comment]
I'm on koreader today's nightly build for android arm64, 2026.03-175-g9c5e59609/

I don't know if this patch is correct but it did change the behavior, now only shift+arrow keys works.
Arrow keys alone still change pages
```lua
-- 2-readerkeyselection-patch.lua

local BD = require("ui/bidi")
local Device = require("device")
local DoubleSpinWidget = require("ui/widget/doublespinwidget")
local Geom = require("ui/geometry")
local InputContainer = require("ui/widget/container/inputcontainer")
local ReaderHighlight = require("apps/reader/modules/readerhighlight")
local Size = require("ui/size")
local SpinWidget = require("ui/widget/spinwidget")
local UIManager = require("ui/uimanager")
local ffiUtil = require("ffi/util")
local logger = require("logger")
local time = require("ui/time")
local _ = require("gettext")
local Screen = Device.screen
local N_ = _.ngettext
local T = ffiUtil.template

-- see https://www.lua.org/gems/sample.pdf for Lua Performance Tips
-- GC Optimisations
local math_abs = math.abs
local math_max = math.max
local math_min = math.min
local math_floor = math.floor

local ReaderKeySelection = require("apps/reader/modules/readerkeyselection")

function ReaderKeySelection:init()
    --     if Device:isTouchDevice() and not Device:hasDPad() then
    --         return
    --     end
    self:registerKeyEvents()
    self.screen_w = Screen:getWidth()
    self.screen_h = Screen:getHeight()
    self._previous_indicator_word = nil
    self._start_indicator_highlight = false
    self._current_indicator_pos = nil
    self._previous_indicator_pos = nil
    self._vertical_move_anchor_x = nil
    self._last_move_was_vertical = false
    self._edge_dx = nil
    self._edge_dy = nil
    self._last_move_was_quick_move = nil
    self.mirroredUI = BD.mirroredUILayout()
    self.ui:registerPostInitCallback(function()
    self.ui.menu:registerToMainMenu(self)
    end)
end
```

**Comment by Commodore64user** [comment]
What are you doing exactly? You don't need a user patch (going by your comment) just need to open the `readerkeyselection` file and comment out those three lines

**Comment by Ceynou** [comment]
I'm repeating the exact same scenario as before
Sorry, I don't understand how I'm supposed to edit a non existing file on android, was I supposed to build the project from source before installing it? I think I forgot how to simply edit it, I'll check it

**Comment by benoit-pierre** [comment]
You can't easily edit those files on Android, unless you have a rooted device.

**Comment by Ceynou** [comment]
I do have a rooted Xiaomi Mi 11i running lineageos 22.1, so I edited the said files with termux and nano, the result is the same effect as the patch apparently.
